File: software/examples/java/ExampleInterrupt.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickletIndustrialDigitalIn4;

public class ExampleInterrupt {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;
	private static final String UID = "XYZ"; // Change to your UID

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickletIndustrialDigitalIn4 idi4 =
		  new BrickletIndustrialDigitalIn4(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Add interrupt listener
		idi4.addInterruptListener(new BrickletIndustrialDigitalIn4.InterruptListener() {
			public void interrupt(int interruptMask, int valueMask) {
				System.out.println("Interrupt Mask: " + Integer.toBinaryString(interruptMask));
				System.out.println("Value Mask: " + Integer.toBinaryString(valueMask));
				System.out.println("");
			}
		});

		// Enable interrupt on pin 0
		idi4.setInterrupt(1 << 0);

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}

File: software/examples/c/example_interrupt.c

#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_industrial_digital_in_4.h"

#define HOST "localhost"
#define PORT 4223
#define UID "xyz" // Change to your UID

// Callback function for interrupts
void cb_interrupt(uint8_t interrupt_mask, uint8_t value_mask) {
	printf("Interrupt by: %d\n", interrupt_mask);
	printf("Value: %d\n", value_mask);
}

int main() {
	// Create IP connectidi4n to brickd
	IPConnection ipcon;
	if(ipcon_create(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not create connectidi4n\n");
		exit(1);
	}

	// Create device object
	IndustrialDigitalIn4 idi4;
	industrial_digital_in_4_create(&idi4, UID); 

	// Add device to IP connectidi4n
	if(ipcon_add_device(&ipcon, &idi4) < 0) {
		fprintf(stderr, "Could not connect to Bricklet\n");
		exit(1);
	}
	// Don't use device before it is added to a connectidi4n

	// Register callback for interrupts
	industrial_digital_in_4_register_callback(&idi4, INDUSTRIAL_DIGITAL_IN_4_CALLBACK_INTERRUPT, cb_interrupt);

	// Enable interrupt on pin 0
	industrial_digital_in_4_set_interrupt(&idi4, 1 << 0);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon);
}
